sql,backupccl: set system table user ID columns to be NOT NULL

andyyang890 opened the pull request:
This PR sets the user ID columns in system tables to be NOT NULL
and when applicable, updates the `RESTORE` logic to account for
the case where a backup may have been created before the user ID
column was added.

Part of #87079

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/98958)
<!-- Reviewable:end -->

andyyang890 committed:
backupccl: update RESTORE logic for system.privileges table

This patch updates the `RESTORE` logic for the `system.privileges`
table to handle restoring a backup that was created before the new
`user_id` column was added.

Release note: None

andyyang890 committed:
backupccl: update RESTORE logic for system.database_role_settings table

This patch updates the `RESTORE` logic for the `system.database_role_settings`
table to handle restoring a backup that was created before the new
`role_id` column was added.

Release note: None

andyyang890 committed:
backupccl: update RESTORE logic for system.external_connections table

This patch updates the `RESTORE` logic for the `system.external_connections`
table to handle restoring a backup that was created before the new
`owner_id` column was added.

Release note: None

andyyang890 committed:
sql: set system table user ID columns to be NOT NULL

This patch sets the user ID columns in system tables to be NOT NULL
and when applicable, updates the migrations that added the columns.

Release note: None

andyyang890 committed:
username: add EmptyRole

This patch adds the EmptyRole, which is used within some system tables.
It also assigns this role the fixed ID of 0.

Release note: None

andyyang890 committed:
sql: fix inserting empty role into system.database_role_settings

Release note: None

andyyang890 committed:
server{,ccl}: update tests to work with system.web_sessions user_id col

Release note: None

rafiss reviewed (approved):
thanks for breaking this down into small commits. it made it very easy to follow. lgtm!

could use a sign-off from the DR team as well.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/98958#-:-NR-TxX14Ox6UPNscpAf:b-hsomy7)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz, @rhu713, and @stevendanna)



<!-- Sent from Reviewable.io -->

stevendanna reviewed (approved):
LGTM! Thanks!

andyyang890 commented:
TFTRs!

bors r=rafiss,stevendanna
